### PR TITLE
12480: GeoAxis Support

### PR DESCRIPTION
**Pull request by bbarden**

### CODE REVIEW

File: bfapi/service/user.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            headers={'Authorization': geo_auth_token},
+        )
+        response.raise_for_status()
+        print(geo_auth_token, response)
+    except requests.ConnectionError:
+        raise CouldNotReachError(userprofile_addr)
+    except requests.HTTPError as err:
+        status_code = err.response.status_code
+        if status_code == 401:
+            raise UnauthorizedError(userprofile_addr)
```
Comment (dbazile) [inline bfapi/service/user.py]: Also, it's probably unnecessary to attach the URL here as the only dynamic part of it is the domain (passed in via ENV) and would likely not provide any added benefit during troubleshooting.

File: bfapi/service/user.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+class Error(Exception):
+    def __init__(self, message: str):
+        super().__init__(message)
+
+class DatabaseResultError(Error):
+    def __init__(self, func_name: str, db_entry: str):
+        super().__init__("Database result error: {} returned entry with incorrect {}.".format(func_name, db_entry))
+
+class CouldNotReachError(Error):
```
Comment (dbazile) [inline bfapi/service/user.py]: Can we rename this to `GeoaxisUnreachable` for clarity?

Ref:
- https://github.com/venicegeo/bf-api/blob/master/bfapi/piazza.py#L642

File: bfapi/service/user.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    def __init__(self, targ_addr: str):
```
Comment (dbazile) [inline bfapi/service/user.py]: It's probably unnecessary to attach the GeoAxis URL here as the only dynamic part of it is the domain (passed in via ENV) and would likely not provide any added benefit during troubleshooting.

File: bfapi/service/user.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        super().__init__('{} is unreachable'.format(targ_addr))
+
+class UnauthorizedError(Error):
+    def __init__(self, targ_addr: str):
```
Comment (dbazile) [inline bfapi/service/user.py]: It's probably unnecessary to attach the GeoAxis URL here as the only dynamic part of it is the domain (passed in via ENV) and would likely not provide any added benefit during troubleshooting.

File: bfapi/service/user.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        super().__init__('Not authorized to access {}'.format(targ_addr))
+
+class HttpCodeError(Error):
+    def __init__(self, status_code: int, targ_addr: str):
+        super().__init__('Http Code error (HTTP {}) at {}'.format(status_code, targ_addr))
+        self.status_code = status_code
+
+class ResponseError(Error):
```
Comment (dbazile) [inline bfapi/service/user.py]: Can we rename this to `InvalidGeoaxisResponse` for clarity?  `ResponseError` in the context of a module called user is fairly ambiguous.

File: bfapi/service/user.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+class CouldNotReachError(Error):
+    def __init__(self, targ_addr: str):
+        super().__init__('{} is unreachable'.format(targ_addr))
+
+class UnauthorizedError(Error):
+    def __init__(self, targ_addr: str):
+        super().__init__('Not authorized to access {}'.format(targ_addr))
+
+class HttpCodeError(Error):
```
Comment (dbazile) [inline bfapi/service/user.py]: Can we rename this to `GeoaxisError` for clarity?  `HttpCodeError` in the context of a module called `user` is fairly ambiguous.

Ref:
- https://github.com/venicegeo/bf-api/blob/master/bfapi/service/scenes.py#L200

File: bfapi/service/user.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    def __init__(self, status_code: int, targ_addr: str):
+        super().__init__('Http Code error (HTTP {}) at {}'.format(status_code, targ_addr))
+        self.status_code = status_code
+
+class ResponseError(Error):
+    def __init__(self, message: str, response_text: str, targ_addr: str):
```
Comment (dbazile) [inline bfapi/service/user.py]: Can you expand the abbreviation `targ_addr` to `target_address` and assign it to `self.target_address`?

File: bfapi/service/user.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    def __init__(self, targ_addr: str):
+        super().__init__('{} is unreachable'.format(targ_addr))
+
+class UnauthorizedError(Error):
+    def __init__(self, targ_addr: str):
+        super().__init__('Not authorized to access {}'.format(targ_addr))
+
+class HttpCodeError(Error):
+    def __init__(self, status_code: int, targ_addr: str):
+        super().__init__('Http Code error (HTTP {}) at {}'.format(status_code, targ_addr))
```
Comment (dbazile) [inline bfapi/service/user.py]: Can you expand the abbreviation `targ_addr` to `target_address`?

File: sql/schema.install.sql [hunk 9]
```diff
@@ -50,12 +50,20 @@ CREATE TABLE __beachfront__detection (
...
+    created_on        TIMESTAMP      NOT NULL    DEFAULT CURRENT_TIMESTAMP
+);
+
 CREATE TABLE __beachfront__job_user (
     job_id            VARCHAR(64),
     user_id           VARCHAR(64),
 
     PRIMARY KEY (job_id, user_id),
     FOREIGN KEY (job_id) REFERENCES __beachfront__job(job_id) ON DELETE CASCADE
+--    FOREIGN KEY (user_id) REFERENCES __beachfront__user(geoaxis_uid) ON DELETE CASCADE
```
Comment (dbazile) [inline sql/schema.install.sql]: Can you uncomment this code and rename `geoaxis_uid` to `user_id` for foreign key consistency?

File: bfapi/config.py [hunk 10]
```diff
@@ -86,6 +86,7 @@ def collect(node: dict):
 PZ_GATEWAY   = os.getenv('PZ_GATEWAY', 'pz-gateway.' + DOMAIN)
 CATALOG      = os.getenv('CATALOG', 'pzsvc-image-catalog.' + DOMAIN)
 TIDE_SERVICE = os.getenv('TIDE_SERVICE', 'bf-tideprediction.' + DOMAIN)
+GEOAXIS_ADDR = os.getenv('GEOAXIS', 'gxisaccess.gxaccess.com')
```
Comment (bbarden) [inline bfapi/config.py]: Done.

File: bfapi/db/user.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+# specific language governing permissions and limitations under the License.
+
+from datetime import datetime
+
+from bfapi.db import Connection, ResultProxy
+
+def select_user(
+        conn: Connection,
+        *,
+        geoaxis_uid: str) -> ResultProxy:
```
Comment (bbarden) [inline bfapi/db/user.py]: discussed and reconsidered.  Appropriate comment made to database (I think?)

File: sql/schema.install.sql (hunk 9 above)
Comment (bbarden) [inline sql/schema.install.sql]: discussed.  Appropriate changes applied

File: test/service/test_users.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        self.mock_get_user_by_api_key.return_value.fetchone.return_value = None
+        with self.assertRaises(users.BeachfrontUnauthorized):
+            new_user = users.authenticate_via_api_key('test-api_key')
+        
+    def test_verifies_api_key(self):
+        self.mock_get_user_by_api_key.return_value.fetchone.return_value = create_user_db_record()
+        with self.assertRaises(users.DatabaseMismatchError):
+            new_user = users.authenticate_via_api_key('not-the-returned-api_key')
+        
+    def test_return_good_user_id(self):
```
Comment (dbazile) [inline test/service/test_users.py]: Can you rename this to `test_assigns_correct_*`?

Ref:
https://github.com/venicegeo/bf-api/blob/master/test/service/test_productlines.py#L412

File: test/service/test_users.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        self.mock_get_user_by_api_key.return_value.fetchone.return_value = create_user_db_record(api_key='test-api_key')
+        new_user = users.authenticate_via_api_key('test-api_key')
+        self.assertIsInstance(new_user, users.User)
+        self.assertEqual('test-user-id', new_user.geoaxis_uid)
+
+    def test_return_good_user_name(self):
```
Comment (dbazile) [inline test/service/test_users.py]: Can you rename this to `test_assigns_correct_*`?

Ref:
https://github.com/venicegeo/bf-api/blob/master/test/service/test_productlines.py#L412

File: test/service/test_users.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+        self.mock_get_user_by_api_key.return_value.fetchone.return_value = create_user_db_record(api_key='test-api_key')
+        new_user = users.authenticate_via_api_key('test-api_key')
+        self.assertIsInstance(new_user, users.User)
+        self.assertEqual('test-user-name', new_user.user_name)
+
+    def test_return_good_api_key(self):
```
Comment (dbazile) [inline test/service/test_users.py]: Can you rename this to `test_assigns_correct_*`?

Ref:
https://github.com/venicegeo/bf-api/blob/master/test/service/test_productlines.py#L412

File: test/service/test_users.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+        new_user = users.authenticate_via_api_key('test-api_key')
+        self.assertIsInstance(new_user, users.User)
+        self.assertEqual('test-api_key', new_user.bf_api_key)
+
+class GetByIdTest(unittest.TestCase):
+    def setUp(self):
+        self._mockdb = helpers.mock_database()
+        self._logger = logging.getLogger('bfapi.service.users')
+        self._logger.disabled = True
+        self.mock_get_user = self.create_mock('bfapi.db.users.select_user')
```
Comment (dbazile) [inline test/service/test_users.py]: Can you change `get` to `select` for consistency?

File: test/service/test_users.py [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+        patcher = patch(target_name)
+        self.addCleanup(patcher.stop)
+        return patcher.start()
+        
+    def test_throws_on_db_error(self):
+        self.mock_get_user.side_effect = helpers.create_database_error()
+        with self.assertRaises(DatabaseError):
+            users.get_by_id('test-user-id')
+        
+    def test_return_none_if_none(self):
```
Comment (dbazile) [inline test/service/test_users.py]: Can you rename to `test_returns_nothing_if_record_not_found`?

File: test/service/test_users.py [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+        self.mock_get_user.return_value.fetchone.return_value = None
+        new_user = users.get_by_id('test-user-id')
+        self.assertEqual(None, new_user)
+        
+    def test_throws_on_wrong_id(self):
+        self.mock_get_user.return_value.fetchone.return_value = create_user_db_record()
+        with self.assertRaises(users.DatabaseMismatchError):
+            users.get_by_id('not-the-returned-geoaxis-id')
+
+    def test_return_good_user_id(self):
```
Comment (dbazile) [inline test/service/test_users.py]: Can you rename this to `test_assigns_correct_*`?

Ref:
https://github.com/venicegeo/bf-api/blob/master/test/service/test_productlines.py#L412

File: test/service/test_users.py [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+        self.mock_get_user.return_value.fetchone.return_value = create_user_db_record()
+        new_user = users.get_by_id('test-user-id')
+        self.assertIsInstance(new_user, users.User)
+        self.assertEqual('test-user-id', new_user.geoaxis_uid)
+
+    def test_return_good_user_name(self):
```
Comment (dbazile) [inline test/service/test_users.py]: Can you rename this to `test_assigns_correct_*`?

Ref:
https://github.com/venicegeo/bf-api/blob/master/test/service/test_productlines.py#L412

File: test/service/test_users.py [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+        self.assertIsInstance(new_user, users.User)
+        self.assertEqual('test-user-name', new_user.user_name)
+
+    def test_return_good_api_key(self):
+        self.mock_get_user.return_value.fetchone.return_value = create_user_db_record()
+        new_user = users.get_by_id('test-user-id')
+        self.assertIsInstance(new_user, users.User)
+        self.assertEqual('test-key', new_user.bf_api_key)
+
+        # TODO? other tests: verify that the calls to the database functions are as they should be.
```
Comment (dbazile) [inline test/service/test_users.py]: Can you replace this comment with one or more stubbed test cases with what you had in mind?

Ref:
https://github.com/venicegeo/bf-api/blob/master/test/db/test_productlines.py#L22-L23

File: bfapi/service/users.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+        log.error('Geoaxis responded to userprofile call without user name.  Response Text: '.format(response.text))
+        raise InvalidGeoaxisResponse('missing `username`', response.text)
+    
+    conn = db.get_connection()
+    try:
+        db.users.insert_or_update_user(
+            conn,
+            geoaxis_uid=geoaxis_uid,
+            user_name=user_name,
+            api_key=uuid.uuid4(),
```
Comment (bbarden) [inline bfapi/service/users.py]: Yes, of course.  Meant to do that the first time around.  Thanks for the catch.

File: test/service/test_users.py [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+from bfapi.db import DatabaseError
+from bfapi.config import GEOAXIS
+
+
+class AuthenticateViaGeoaxisTest(unittest.TestCase):
+    def setUp(self):
+        self._logger = logging.getLogger('bfapi.service.users')
+        self._logger.disabled = True
+        self._mockdb = helpers.mock_database()
+        self._logger = logging.getLogger('bfapi.service.jobs')
```
Comment (bbarden) [inline test/service/test_users.py]: Argh.  Thanks for the catch.

File: test/service/test_users.py [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+        with self.assertRaises(users.GeoaxisUnauthorized):
+            users.authenticate_via_geoaxis('test-token')
+
+    def test_throws_on_http_error(self):
+        self.mock_requests.get(self.userprofile_addr, text='resp', status_code=405)
+        with self.assertRaises(users.GeoaxisError):
+            users.authenticate_via_geoaxis('test-token')
+        
+    def test_throws_on_no_uid(self):
+        self.mock_requests.get(self.userprofile_addr, text='{"username":"test-name"}', status_code=201)
```
Comment (bbarden) [inline test/service/test_users.py]: To the best of my knowledge it is.  It doesn't make `response.raise_for_status()` throw an error, whereas `status_code=401` and `status_cod=405` do.  I spent a while banging at this stuff to get it to the point where it seemed to work right, and I *think* I'm doing it the correct way.

File: test/service/test_users.py [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+        with self.assertRaises(users.InvalidGeoaxisResponse):
+            users.authenticate_via_geoaxis('test-token')
+        
+    def test_throws_on_no_user_name(self):
+        self.mock_requests.get(self.userprofile_addr, text='{"uid":"test-id"}', status_code=201)
```
Comment (bbarden) [inline test/service/test_users.py]: as above.

File: test/service/test_users.py [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+        with self.assertRaises(users.InvalidGeoaxisResponse):
+            users.authenticate_via_geoaxis('test-token')
+        
+    def test_throws_on_dbase_error(self):
+        self.mock_requests.get(self.userprofile_addr, text='{"uid":"test-id","username":"test-name"}', status_code=201)
```
Comment (bbarden) [inline test/service/test_users.py]: as above

File: test/service/test_users.py (hunk 12 above)
Comment (bbarden) [inline test/service/test_users.py]: I think that may be overselling the test a bit, but I've made the change.

File: test/service/test_users.py (hunk 13 above)
Comment (bbarden) [inline test/service/test_users.py]: as above

File: test/service/test_users.py (hunk 14 above)
Comment (bbarden) [inline test/service/test_users.py]: as above

File: test/service/test_users.py (hunk 19 above)
Comment (dbazile) [inline test/service/test_users.py]: Examples of these are:

1. Checking that a call to the `bfapi.db.users.do_something` happened: https://github.com/venicegeo/bf-api/blob/eb7ad4963ccf43536d7d111c20c03366c0a52dc3/test/service/test_jobs.py#L228-L238

2. Checking the call parameters for a call to `bfapi.db.users.do_something`: https://github.com/venicegeo/bf-api/blob/3c1be3515a953d76c85b963562d40f53b5ab0eed/test/routes/test_init.py#L57

File: test/service/test_users.py (hunk 22 above)
Comment (dbazile) [inline test/service/test_users.py]: For this particular test case, you're checking that `authenticate_via_geoaxis` verifies the response payload which implies the request did not have an error status code.  The status code for this particular mocked call is returning an `HTTP 201 Created` which is likely not the response code GeoAxis returns for a "get profile" request.

Comment (bbarden) [inline test/service/test_users.py]: Changed to 200.

Comment (dbazile) [inline test/service/test_users.py]: :+1: 

For future reference, the `status_code` parameter defaults to `200` and can be omitted if that's the case.

**Pull request merged**